apps/passport-client/components/screens/LoginScreens/LoginScreen.tsx

```diff
@@ -126,6 +130,93 @@ export function LoginScreen(): JSX.Element {
     [dispatch, email]
   );
 
+  const [tryStorageAccess, setTryStorageAccess] = useState(false);
+  const [tryStorageLogin, setTryStorageLogin] = useState(false);
+
+  /**
+   * Assuming we're in Chrome and an iframe, and we've successfully loaded an
+   * encryption key from local storage, try to use it to log in.
+   */
+  const tryToLogin = useCallback(
+    async (encryptionKey: string) => {
+      // Try to download and decrypt the storage
+      const storageRequest = await requestDownloadAndDecryptStorage(
+        appConfig.zupassServer,
+        encryptionKey
+      );
+      if (storageRequest.success) {
+        // Success, log in
+        dispatch({
+          type: "load-after-login",
+          storage: storageRequest.value,
+          encryptionKey
+        });
+      }
+    },
+    [dispatch]
+  );
+
+  /**
+   * This will only be called if we're in an iframe and Chrome.
+   */
+  const requestStorageAndLogIn = useCallback(async () => {
+    try {
+      // @ts-expect-error Chrome-only API
+      const handle: { localStorage: Storage } =
+        // @ts-expect-error Chrome-only API
+        await document.requestStorageAccess({ localStorage: true });
+
+      // Access granted, try reading the local storage
+      const encryptionKey = handle.localStorage.getItem("encryption_key");
```

Comment:
It would be nice if this could just go through a normal load using the existing local storage rather than doing an extra download + login and creating a separate "device" for this iframe.  The separate local storage is going to create more opportunities for sync conflicts and stale data.

That being said, it looks like doing that might be complicated, since there are multiple things to request access to beyond local storage, and I'm not sure which of this will cause something to unexpectedly break in future (e.g. broadcast channels for talking to the service worker, or shared workers if we use them in future).  This approach seems sound for now.

Reply:
I think we could revisit this if it seems like it's a necessary optimization. It might be useful to cut down on the amount of sync traffic - I'll add a Linear issue for this.

Reply:
It would avoid a download, but the sync traffic after this point would be the same.
It would importantly also reduce the number of copies of Zupass state stored on the user's device, which could be a size concern.
If there's a future version of sync which knows how to load and merge from LocalStorage then there would be a bigger difference in behavior, but that's theoretical.
None of these are important concerns for now.